internal/agent/device/image/image.go

```diff
@@ -0,0 +1,104 @@
+package image
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/flightctl/flightctl/api/v1alpha1"
+	"github.com/flightctl/flightctl/internal/container"
+)
+
+type Image struct {
+	Base   string
+	Tag    string
+	Digest string
+}
+
+// Returns the image formatted in a <base>:<tag>@<digest> format
+func (i Image) String() string {
+	if i.Digest != "" {
+		if i.Tag != "" {
+			return fmt.Sprintf("%s:%s@%s", i.Base, i.Tag, i.Digest)
+		}
+		return fmt.Sprintf("%s@%s", i.Base, i.Digest)
+	}
+	if i.Tag != "" {
+		return fmt.Sprintf("%s:%s", i.Base, i.Tag)
+	}
+	return i.Base
+}
+
+// Bootc does not accept images with a tag AND digest specified
+// In cases when both are present the digest will be returned and NOT a tag
+func (i *Image) ToBootcTarget() string {
+	if i.Digest != "" {
+		return fmt.Sprintf("%s@%s", i.Base, i.Digest)
+	}
+	if i.Tag != "" {
+		return fmt.Sprintf("%s:%s", i.Base, i.Tag)
+	}
+	return i.Base
+}
+
+func parseImage(image string) *Image {
+	imageObj := &Image{}
+
+	imageAndDigest := strings.SplitN(image, "@", 2)
+	if len(imageAndDigest) == 2 {
+		imageObj.Digest = imageAndDigest[1]
+	}
+
+	imageAndTag := strings.SplitN(imageAndDigest[0], ":", 2)
+	imageObj.Base = imageAndTag[0]
+	if len(imageAndTag) == 2 {
+		imageObj.Tag = imageAndTag[1]
+	}
+
+	return imageObj
+}
+
+func SpecToImage(spec *v1alpha1.DeviceOSSpec) *Image {
+	if spec == nil || spec.Image == "" {
+		return nil
+	}
+	return parseImage(spec.Image)
+}
+
+func BootcStatusToImage(bootc *container.BootcHost) *Image {
+	if bootc == nil {
```

Comment:
In general, if a function returns nil without a non nil error we should be asking why. This ends up exposing lots of other things to the possibility of nil pointer. Go code expects that nil checks are done before you pass anything to a function that is maintained by a very boring pattern.

```
something, err := myFunc()
if err != nil {
  return err
 }
 ```
 here we can assume if the function return a non nil error that the returned value is not nil. This is repetitive and maybe ugly but it provides us with guardrails and reasonable assumptions.
 
 here I would argue that you should never be able to create bootc that is nil. If you can then that should get fixed.
 
 Of course, there are exceptions when we are dealing with serialization etc I think you get what I am saying :)